packages/types/es/src/produce.ts

```diff
@@ -0,0 +1,76 @@
+/*!
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+/* eslint-disable @typescript-eslint/no-explicit-any */
+
+import { IMapProducer, IProducer, IShapeProducer, IVectorProducer } from "@tiny-calc/types";
+
+// Assigned to IProducer.close(), serves as a sentinel to detect and break cycles.
+const unitFn = () => {};
+
+const props = {
+    // #region IProducer
+    open:  { value: function() { return this; }},
+    close: { value: unitFn },
+    // #endregion IProducer
+
+    // #region IReader
+    get:   { value: function(key: PropertyKey) { return (this as any)[key]; }},
+    // #endregion IReader
+
+    // #region IShapeReader
+    keys:  { value: function() { return Object.keys(this); }},
+    has:   { value: function (key: any) { return Object.keys(this).includes(key); }},
+    size:  { get: function() { return Object.keys(this).length; }},
+    // #endregion IShapeReader
+}
+
+const vectorProps = {
+    // #region IProducer, IReader, IShapeReader
+    ...props,
```

Comment:
Is this approach faster than a class? Looks clean though.

Reply:
Yeah, I feel dirty about it...  but the perf is excellent:
* For map-like objects, reader.get("bar") is ~ as fast as reader["bar"]
* For arrays, reader.getItem(43) is ~ as fast as reader[43]

Probably the most common usage of 'produce' will be combined with JSON parse, in which case we can wrap it up like this so that no one can get their hands on the mutable intermediate objects:
```ts
const producer: IProducer = Producer.parse("{ a: 1, b: 2 }");
```
...or we can have "debug" builds where produce(..) implicitly freezes the source objects to help make this pattern safer.